main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala

```diff
@@ -281,6 +337,115 @@ object Interpreter {
     case Expression.Universal(params, exp, loc) => throw InternalRuntimeException(s"Unexpected expression: '$exp' at ${loc.source.format}.")
   }
 
+  //
+  // Get value from channel.
+  //
+  private def getChannel(chan: Value.Channel): AnyRef = {
+    var value: AnyRef = null
+
+    chan.cLock.lock()
+    try {
+      while (chan.queue.isEmpty)
+        chan.bufferNotFull.await()
+
+      assert(!chan.queue.isEmpty)
+
+      value = chan.queue.poll()
+      if (value != null)
+        chan.bufferNotEmpty.signalAll()
+    }
+    finally
+      chan.cLock.unlock()
+
+    value
+  }
+
+  //
+  // Put value to channel.
+  //
+  private def putChannel(chan: Value.Channel, value: AnyRef): AnyRef = {
+    chan.cLock.lock()
+    try {
+      while (chan.queue.size() == chan.capacity)
+        chan.bufferNotEmpty.await()
+
+      assert(chan.queue.size() != chan.capacity)
+
+      chan.queue.add(value)
+      chan.bufferNotFull.signalAll()
+
+      signalConditions(chan.conditions)
+      chan.conditions.clear()
+    }
+    finally
+      chan.cLock.unlock()
+
+    chan
+  }
+
+  private def getSelect(rules: List[(Symbol.VarSym, Value.Channel, Expression)]): (Symbol.VarSym, AnyRef, Expression) = {
+    val sLock = new ReentrantLock
+    val sCondition = sLock.newCondition
+    val channels = rules.map(_._2)
+
+    var result: (Symbol.VarSym, AnyRef, Expression) = null
+
+    sLock.lock()
+    lockChannels(channels)
+    try {
+      while (result == null) {
+        result = pollChannels(rules)
+
+        if (result == null) {
+          addConditions(channels, sLock, sCondition)
+          unlockChannels(channels)
+          sCondition.await()
+          lockChannels(channels)
+        }
+      }
+    }
+    finally {
+      sLock.unlock()
+      unlockChannels(channels)
+    }
+
+    result
+  }
+
+  private def lockChannels(channels: List[Value.Channel]) =
+    channels.foreach(_.cLock.lock())
```

Comment:
We usually use a normal for loop to better highlight something which is nonfunctional like foreach.